kartothek/io/iter.py

```diff
@@ -27,6 +30,36 @@
 )
 
 
+class StorePool:
+    def __init__(self, store_factory):
+        """Generally we cannot assume that store objects are thread safe. Therefore, pool the objects to limit the number of required connections"""
+        # deques are faster than ordinary queues since they don't require locking
+        self._pool = deque()
+        self.store_factory = store_factory
+
+    def store(self):
+        try:
```

Comment:
Use `len(self._pool)` instead of  exceptions

Reply:
Wouldn't guarantee that the pop will be successful, would it?

Reply:
You would need a mutex around this section obviously.

Reply:
I'll look into the solution with mutex but my first try included a queue and the sync/lock overhead was quite significant